design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,255 @@ status: draft
+
+## Delegates
+
+Delegates are xlang types that act as a type-safe function pointer. They are similar conceptually
+to an [interface](#interfaces) that declares one and only one method. Defining delegates as a
+separate type from interfaces enables better integration with free functions and/or [anonymous
+functions](https://en.wikipedia.org/wiki/Anonymous_function) in language projections. Delegates are
+often (but not exclusively) used to declare the signature for xlang event handlers.
+
+xlang delegates are named types and define a method signature. Delegate method signatures follow the
+same rules for parameters as interface methods do.
+
+Like interfaces, delegates require a GUID identifier. THis identifier can be provided explicitly or
+generated implicitly from the type's name.
+
+Like interfaces, delegates can be parameterized. Parameterized delegates may declare arguments
+and/or their return value type in terms of the type parameters instead of as concrete types.
+
+Delegates are not versionable. Once they have been defined, they may never be changed.
+
+## Classes
+
+Classes are xlang types that expose behavior and encapsulate data. Often times, classes can be activated
+
+Classes are additively versionable. Subsequent versions of a given enum may add new members. 
+Pre-existing members may not be removed or modified in any way.
+
+## Object Members
+
+### Methods
+
+#### Method Overloading
+
+#### Parameters
+
+##### Array Parameters
+
+### Properties
+
+### Events
+
+### Intrinsic Operations
+
+All interfaces and runtime classes support the following operations intrinsically. The projection of
+these intrinsic operations is language specific. For languages that have a standard projection for
+a given intrinsic operation (such as C++ ToString or JavaScript equals), the projection is free to
+project the operation in an appropriate manner, if at all.
+
+#### ToString
+
+Returns a string that represents the current object. Typically, this defaults to the object's fully
+namespace-qualified type name, but can be customized. For example, an xlang type that represents a
+[JSON value](http://json.org/) would return the text representation of the JSON value from ToString.
+
+Projection language examples:
+
+- C++ N/A
+- [Java Object.toString](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#toString())
+- [JavaScript Object.prototype.toString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString)
+- [.NET Object.ToString](https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netframework-4.7.2)
+- [Objective-C NSObject description](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418746-description)
+- [Python __str__ magic method](https://docs.python.org/3/library/stdtypes.html#str)
+
+#### GetHashCode
+
+Returns an Int32 hash code that can be used to insert and identify an object in a hash-based collection.
```

Comment:
Today language projections produce the hash and can thus produce pointer-sized hash values for efficiency. This seems like a limitation.

Reply:
Language projections can't provide value semantics for hash codes or equals - hence the reason for providing types the ability to customize those values. 

I was thinking we should provide a way to indicate (via custom attribute) "this type has reference semantics" as a hint to the language projection NOT to bother calling gethashcode or equals.

Reply:
As an aside, we really need to support IntPtr. So many WinRT APIs have hacky workarounds for this limitation.

Reply:
Good idea. Can you provide some examples from WinRT so I can review?

Reply:
Event tokens for one. :) This is also the main reason WinRT APIs include "interop" headers that are manually written and shipped in the SDK, for example `windows.graphics.capture.interop.h`. There are some others where API authors smuggle "handles" as UINT64 values. There are legitimate reasons for this limitation though when it comes to marshaling across bitness and nothing inherently wrong with saying "just use UINT64". What you can't say is "just use UINT32".

Reply:
Good point about multiple language projections @BrentRe

Reply:
@zooba I am 100% *against* enabling pointer arithmetic for whatever we end up callign IntPtr

Reply:
I'm not really arguing one way or the other, but this is how it works today in WinRT and there have been no complaints that I'm aware of, but I'm curious how exactly you would "pass a hash table from one language to another"? In WinRT today we pass *implementations* of hash tables around and the implementation controls the hashing algorithm. This is unaffected by other languages that may happen to be interacting with the data structure via some interface like `IMap<K,V>`.

Reply:
Take, for example, a map of Point to String. One language projection of Point adds the Point to the map. That language projection calculates the hash code of Point by shifting X left 8 bits and adding it to Y.

Now you pass the collection to a different language projection. Code in that language attempts to retrieve a Map entry. It creates its own projection of a Point with the same X and Y values. But this language projection calculates the hash code differently - crypto hash, or shifting Y instead of X, or whatever. 

Now that language projection will fail to locate the Point in the Map.

Reply:
Yes, I know that's what you meant but I'm saying that's not how it works. When I call `IMap<K, V>::Find` I don't provide a hash code. I simply provide the key. The implementation (that knows about the implementation) performs the hash operation as needed on the key it receives. The caller doesn't even know whether a hash is needed. The IMap could be implemented using an ordered map algorithm (that doesn't require a hash value).

Reply:
How does a hypothetical implementation of Map, one that needs a hash code, get a hash code for an opaque object used as a key?

Won't it need to get it from the opaque object?

If so, I'm simply observing that if we decide that a projection provide the hash code, rather than the opaque object itself, then containers holding things originating from different projections will have problems. We can avoid these problems by requiring the opaque object itself to supply the hash code.

In .NET, for example, the Map doesn't compute the hash code for the key being added to the Map. Rather the Map asks the key object for its hash code.

Reply:
An implementation of IMap can do whatever it wants. It can calculate the hash code from the raw ABI pointer for all I care.

I'm having more and more doubts about introducing something fundamental to the type system like hashability, equality, and sort ordering, because the "one size fits all" approach is bound to hit limitations. I think these should be implementation details, specific to the scenario that the container implementer wants to optimize for.

Reply:
It's not really a question of what we think is needed so much as whan the languages we project into will support. Java and .NET's intrinsic hash code type is int32. ObjectiveC and Python use a pointer sized signed integer (i.e. IntPtr). C++ uses std::size_t, which is unsigned and implementation dependent (but typically also pointer sized as I understand).

So the real question is if we want to limit the size to 32 bit because of Java and .NET or make it 64 bit and force those projections do adapt the 64 bit hash to 32 bit. 

I will add an open issue to capture this